packages/core/src/indexing/client.ts

```diff
@@ -0,0 +1,998 @@
+      return async (...args: Parameters<typeof action>) => {
+        const RETRY_COUNT = 3;
+        for (let i = 0; i <= RETRY_COUNT; i++) {
+          try {
+            // @ts-ignore
+            return await action(...args);
+          } catch (error) {
+            if (
+              (error as Error)?.message?.includes("returned no data") ===
+                false ||
+              i === RETRY_COUNT
+            ) {
+              throw error;
+            }
+          }
+        }
+      };
+    };
+
+    for (const action of blockDependentActions) {
+      actions[action] = getPonderAction(action);
+    }
+
+    // @ts-ignore
+    actions.multicall = getRetryAction(getPonderAction("multicall"));
+    // @ts-ignore
+    actions.readContract = getRetryAction(getPonderAction("readContract"));
+    actions.simulateContract = getRetryAction(
+      // @ts-ignore
+      getPonderAction("simulateContract"),
+    );
+
+    for (const action of nonBlockDependentActions) {
+      // @ts-ignore
+      actions[action] = _publicActions[action];
+    }
+
+    // required block actions
+
+    for (const action of [
+      "getBlock",
+      "getBlockTransactionCount",
+      "getTransactionCount",
+    ] as const) {
+      // @ts-ignore
+      actions[action] = _publicActions[action];
+    }
+
+    const actionsWithMetrics = {} as PonderActions;
+
+    for (const [action, actionFn] of Object.entries(actions)) {
+      // @ts-ignore
+      actionsWithMetrics[action] = async (
+        ...args: Parameters<PonderActions[keyof PonderActions]>
+      ) => {
+        const endClock = startClock();
+        try {
+          // @ts-ignore
+          return await actionFn(...args);
+        } finally {
+          common.metrics.ponder_indexing_rpc_action_duration.observe(
+            { action },
+            endClock(),
+          );
+        }
+      };
+    }
+
+    return actionsWithMetrics;
+  };
+
+  return {
+    getClient(network) {
+      const requestQueue =
+        requestQueues[indexingBuild.networks.findIndex((n) => n === network)]!;
+
+      return createClient({
+        transport: cachedTransport({
+          common,
+          network,
+          requestQueue,
+          syncStore,
+          cache,
+        }),
+        chain: network.chain,
+        // @ts-expect-error overriding `readContract` is not supported by viem
+      }).extend(ponderActions);
+    },
+    async prefetch({ events, eventCount }) {
+      // Use profiling metadata + next event batch to determine which
+      // rpc requests are going to be made, and preload them into the cache.
+
+      const prediction: { ev: number; request: Request }[] = [];
+
+      for (const event of events) {
+        if (profile.has(event.name)) {
+          for (const [, { pattern, count }] of profile.get(event.name)!) {
+            // Expected value of times the prediction will be used.
+            const ev = count / eventCount[event.name]!;
```

Comment:
Are these are both total counts since the beginning of indexing?

Reply:
yes